File: packages/asdl-objectives/src/asdl_objectives/list_branch_attribution.py
"""Local branch attribution for ``objective list`` records."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Final

from asdl_core.git.git_gateway import GitGateway
from asdl_core.git.types import GitCommandFailure, LocalBranchTip, PathChangeTouch
from asdl_objectives.list_updates import parse_iso_datetime
from asdl_objectives.objective_storage import (
    active_root_relative_path,
    objective_slug_from_active_path,
)

MAX_UPDATED_BRANCH_ATTRIBUTION_WALKS: Final = 50


@dataclass(frozen=True)
class ObjectiveBranchAttribution:
    updated_branches_by_slug: dict[str, tuple[str, ...]]
    truncated: bool = False


def build_objective_branch_attribution(
    git: GitGateway,
    *,
    trunk_branch: str,
    slugs: frozenset[str],
    max_branch_walks: int = MAX_UPDATED_BRANCH_ATTRIBUTION_WALKS,
) -> ObjectiveBranchAttribution | GitCommandFailure:
    if not slugs:
        return ObjectiveBranchAttribution(updated_branches_by_slug={})

    updated_branches_by_slug: dict[str, list[str]] = {slug: [] for slug in slugs}
    branches = _local_non_trunk_branches(git, trunk_branch=trunk_branch)
    if not branches:
        return _attribution_from_lists(updated_branches_by_slug)

    objective_root = active_root_relative_path().as_posix()
    changed_branches = _branches_with_objective_tree_changes(
        git,
        trunk_branch=trunk_branch,
        branches=branches,
        objective_root=objective_root,
    )
    if isinstance(changed_branches, GitCommandFailure):
        return changed_branches

    walked_branches = changed_branches[:max_branch_walks]
    truncated = len(changed_branches) > max_branch_walks
    for branch in walked_branches:
        touches = git.path_touches_under(f"{trunk_branch}..{branch}", objective_root)
        if isinstance(touches, GitCommandFailure):
            return touches

        for slug in sorted(_objective_slugs_from_touches(touches, slugs)):
            updated_branches_by_slug[slug].append(branch)

    return _attribution_from_lists(updated_branches_by_slug, truncated=truncated)


def _attribution_from_lists(
    updated_branches_by_slug: dict[str, list[str]],
    *,
    truncated: bool = False,
) -> ObjectiveBranchAttribution:
    return ObjectiveBranchAttribution(
        updated_branches_by_slug={
            slug: tuple(branch_names) for slug, branch_names in updated_branches_by_slug.items()
        },
        truncated=truncated,
    )


def _objective_slugs_from_touches(
    touches: tuple[PathChangeTouch, ...],
    slugs: frozenset[str],
) -> set[str]:
    touched_slugs: set[str] = set()
    for touch in touches:
        for path in touch.paths:
            slug = objective_slug_from_active_path(path)
            if slug in slugs:
                touched_slugs.add(slug)
    return touched_slugs


def _local_non_trunk_branches(
    git: GitGateway,
    *,
    trunk_branch: str,
) -> tuple[str, ...]:
    branch_tips = sorted(
        (tip for tip in git.list_local_branch_tips() if tip.name != trunk_branch),
        key=_branch_tip_sort_key,
    )
    return tuple(tip.name for tip in branch_tips)


def _branch_tip_sort_key(tip: LocalBranchTip) -> tuple[int, float, str]:
    if tip.head_iso is None:
        return (1, 0.0, tip.name)
    parsed = parse_iso_datetime(tip.head_iso)
    if parsed is None:
        return (1, 0.0, tip.name)
    return (0, -parsed.timestamp(), tip.name)


def _branches_with_objective_tree_changes(
    git: GitGateway,
    *,
    trunk_branch: str,
    branches: tuple[str, ...],
    objective_root: str,
) -> tuple[str, ...] | GitCommandFailure:
    refs = (trunk_branch, *branches)
    tree_oids = git.tree_oids_at_refs(refs, objective_root)
    if isinstance(tree_oids, GitCommandFailure):
        return tree_oids

    trunk_tree_oid = tree_oids.get(trunk_branch)
    return tuple(branch for branch in branches if tree_oids.get(branch) != trunk_tree_oid)

File: ts/packages/objective/src/operations/list-branch-attribution.ts
import type { GitErrorInfo, GitGateway, GitLocalBranchTip } from "@asdl/core/git";

import { activeRootRelativePath, objectiveSlugFromActivePath } from "../storage.ts";

export const MAX_UPDATED_BRANCH_ATTRIBUTION_WALKS = 50;

export interface ObjectiveBranchAttribution {
	updatedBranchesBySlug: ReadonlyMap<string, readonly string[]>;
	isTruncated: boolean;
}

export interface BuildObjectiveBranchAttributionParams {
	repoRoot: string;
	trunkBranch: string;
	slugs: ReadonlySet<string>;
	maxBranchWalks?: number | undefined;
}

export async function buildObjectiveBranchAttribution(
	git: GitGateway,
	params: BuildObjectiveBranchAttributionParams,
): Promise<{ type: "ok"; value: ObjectiveBranchAttribution } | { type: "git-error"; error: GitErrorInfo }> {
	if (params.slugs.size === 0) return { type: "ok", value: emptyAttribution(params.slugs) };

	const tips = await git.listLocalBranchTips({ cwd: params.repoRoot });
	if (!tips.ok) return { type: "git-error", error: tips.error };

	const branches = tips.value.filter((tip) => tip.name !== params.trunkBranch).sort(compareBranchTips).map((tip) => tip.name);
	if (branches.length === 0) return { type: "ok", value: emptyAttribution(params.slugs) };

	const objectiveRoot = activeRootRelativePath();
	const treeOids = await git.treeOidsAtRefs({ cwd: params.repoRoot, refs: [params.trunkBranch, ...branches], relativePath: objectiveRoot });
	if (!treeOids.ok) return { type: "git-error", error: treeOids.error };

	const trunkTreeOid = treeOids.value[params.trunkBranch] ?? null;
	const changedBranches = branches.filter((branch) => (treeOids.value[branch] ?? null) !== trunkTreeOid);
	const maxBranchWalks = params.maxBranchWalks ?? MAX_UPDATED_BRANCH_ATTRIBUTION_WALKS;
	const walkedBranches = changedBranches.slice(0, maxBranchWalks);
	const bySlug = new Map<string, string[]>([...params.slugs].map((slug) => [slug, []]));

	for (const branch of walkedBranches) {
		const changedPaths = await git.changedPathsUnder({ cwd: params.repoRoot, revisionRange: `${params.trunkBranch}...${branch}`, relativePath: objectiveRoot });
		if (!changedPaths.ok) return { type: "git-error", error: changedPaths.error };

		for (const slug of objectiveSlugsFromPaths(changedPaths.value, params.slugs)) {
			bySlug.get(slug)?.push(branch);
		}
	}

	return {
		type: "ok",
		value: {
			updatedBranchesBySlug: freezeAttributionMap(bySlug),
			isTruncated: changedBranches.length > maxBranchWalks,
		},
	};
}

function emptyAttribution(slugs: ReadonlySet<string>): ObjectiveBranchAttribution {
	return { updatedBranchesBySlug: new Map([...slugs].map((slug) => [slug, []])), isTruncated: false };
}

function objectiveSlugsFromPaths(paths: readonly string[], slugs: ReadonlySet<string>): string[] {
	const touchedSlugs = new Set<string>();
	for (const path of paths) {
		const slug = objectiveSlugFromActivePath(path);
		if (slug !== null && slugs.has(slug)) touchedSlugs.add(slug);
	}
	return [...touchedSlugs].sort((left, right) => left.localeCompare(right));
}

function compareBranchTips(left: GitLocalBranchTip, right: GitLocalBranchTip): number {
	const leftTime = parsedTime(left.headIso);
	const rightTime = parsedTime(right.headIso);
	if (leftTime !== null && rightTime !== null && leftTime !== rightTime) return rightTime - leftTime;
	if (leftTime !== null && rightTime === null) return -1;
	if (leftTime === null && rightTime !== null) return 1;
	return left.name.localeCompare(right.name);
}

function parsedTime(iso: string | null): number | null {
	if (iso === null) return null;
	const parsed = Date.parse(iso);
	return Number.isFinite(parsed) ? parsed : null;
}

function freezeAttributionMap(bySlug: ReadonlyMap<string, readonly string[]>): ReadonlyMap<string, readonly string[]> {
	return new Map([...bySlug.entries()].map(([slug, branches]) => [slug, [...branches]]));
}
